sysl modules GitHub

ChloePlanet opened the pull request:
## Description:
### GitHub API based sysl modules
```
$ export SYSL_MODULES=github
$ export SYSL_GITHUB_TOKEN=**************
```
* `import //github.com/anz-bank/sysl-go@v0.0.1` downloads the v0.0.1 of sysl-go
* `import //github.com/anz-bank/sysl-go@master` downloads the up-to-date master branch sysl-go every time
* `import //github.com/anz-bank/sysl-go` simplified statement for `import //github.com/anz-bank/sysl-go@master`

There's no dependency list and lock file required under the working directory. Which means if there are the same importing statements in multiple files, developers need to update all of them when they want to upgrade all of them.
This is not a problem for sysl-catalog for now as sysl-catalog is supposed to import every dependent file once (so only in one file). If the dependency maintaining becomes an issue in the future, one approach is to add an optional lock file, and change the behaviour of `import //github.com/anz-bank/sysl-go` that it matches the version indicating in lock file instead of fetching the latest version every time.

#### Note
The behaviour of go modules based sysl modules is a bit different from GitHub API based sysl modules:
```
$ export SYSL_MODULES=gomod // or SYSL_MODULES=on
```
* `import //github.com/anz-bank/sysl-go@v0.0.1` downloads the v0.0.1 of sysl-go
* `import //github.com/anz-bank/sysl-go@master` downloads the up-to-date master branch sysl-go every time
* `import //github.com/anz-bank/sysl-go` downloads and locks the package version the first time importing sysl-go. Going to add a `sysl mod get -u github.com/anz-bank/sysl-go`/`sysl mod get github.com/anz-bank/sysl-go@v0.0.2` to support update the locked dependency version in a separate PR.

## Changes:
- Added `DependencyManager` interface
  - Implemented GitHub dependency manager which downloads and saves dependent files in `$HOME/.sysl` under GitHub Mode
  - Refactored go modules manager which is the default Sysl Modules dependency manager
- Added tests respectively

## Checklist:
- [x] Added related tests
- [x] Made corresponding changes to the documentation

cuminandpaprika commented on pkg/mod/module.go:
```diff
@@ -8,18 +8,32 @@ import (
 var modules Modules
 
+type DependencyManager interface {
+	Get(filename, ver string, m *Modules) (*Module, error)
```
Some comments would be super nice here, to help distinguish between the Get and Find interface. Looks to me like they're the same except for the returned error?

cuminandpaprika commented on pkg/mod/module_test.go:
```diff
@@ -54,6 +80,15 @@ func TestFindWithWrongPath(t *testing.T) {
 	assert.Nil(t, mod)
 }
+
+func TestHasPathPrefix(t *testing.T) {
+	assert.True(t, hasPathPrefix("github.com/anz-bank/sysl", "github.com/anz-bank/sysl/deps.sysl"))
+	assert.True(t, hasPathPrefix("github.com", "github.com/anz-bank/sysl"))
+	assert.True(t, hasPathPrefix("github.com/", "github.com/anz-bank/sysl"))
+	assert.True(t, hasPathPrefix("github.com/anz-bank/sysl", "github.com/anz-bank/sysl"))
+	assert.True(t, hasPathPrefix("github.com/anz-bank/sysl", "github.com/anz-bank/sysl/"))
+	assert.False(t, hasPathPrefix("github.com/anz-bank/sysl2", "github.com/anz-bank/sysl/deps.sysl"))
+}
```
nit: This might be improved by formatting it as a table driven test, which would allow you to give a name to each test. Totally optional though, current formatting is also okay, I had to read a few times to make sure likes 89-92 weren't duplicates

cuminandpaprika commented on pkg/mod/github.go:
```diff
@@ -0,0 +1,140 @@
+	"github.com/pkg/errors"
+)
+
+var client *github.Client
```
Does `client` have to be a global variable or could it be better encapsulated inside `githubMgr`

ChloePlanet commented on pkg/mod/github.go:
```diff
@@ -0,0 +1,140 @@
+		if err != nil {
+			log.Fatal(err)
+		}
+		syslModulesCacheDir = filepath.Join(usr.HomeDir, ".sysl")
```
Yeah sure

ChloePlanet commented on pkg/mod/github.go:
```diff
@@ -0,0 +1,140 @@
+		ver = "v0.0.0-" + fileContent.GetSHA()[:12]
+	}
+
+	name := filepath.Join("github.com", owner, repo)
+	dir := filepath.Join(syslModulesCacheDir, name)
```
Yes on Windows an example Module would be like
```
mod := &Module{
		Name:    "github.com/anz-bank/sysl",
		Dir:     "C:\Users\foo\go\pkg\mod\github.com\anz-bank\sysl@v1.1.0",
		Version: "1.1.0",
}
```

ChloePlanet commented on pkg/mod/github.go:
```diff
@@ -0,0 +1,140 @@
+}
+
+func (*githubMgr) Get(filename, ver string, m *Modules) (*Module, error) {
+	names := strings.Split(filename, "/")
```
The filename is joined by forward slash `/` in any OS but `filepath.SplitList` splits a list of paths joined by the OS-specific ListSeparator.

ChloePlanet commented:
> I noticed that authentication for the Github API hasn't been addressed yet. Is that to be addressed in a later PR?
I don't want to use my personal access token in a non-personal repo. Maybe after we added a sysl bot account in sysl group.

ChloePlanet commented:
> I don't want to use my personal access token in a non-personal repo. Maybe after we added a sysl bot account in sysl group.

Added a sysl bot (https://github.com/SyslBot) to sysl with read access and use the personal access token of sysl bot in the pr.

cuminandpaprika commented on pkg/mod/module_test.go:
```diff
@@ -85,10 +83,21 @@ func TestFindWithWrongPath(t *testing.T) {
-	assert.True(t, hasPathPrefix("github.com/", "github.com/anz-bank/sysl"))
-	assert.True(t, hasPathPrefix("github.com/anz-bank/sysl", "github.com/anz-bank/sysl"))
-	assert.True(t, hasPathPrefix("github.com/anz-bank/sysl", "github.com/anz-bank/sysl/"))
+	t.Parallel()
```
Awesome! looks much cleaner! 👍

anzdaddy commented:
The rate limiting could be a showstopper if it hits the API too hard. Is this likely to be a problem with the current caching approach? What's involved in using an API key and what are the downsides?

anzdaddy commented:
Do we need per-import versioning? I thought all the use cases we discussed could be solved by other means.

ChloePlanet commented:
> The rate limiting could be a showstopper if it hits the API too hard. Is this likely to be a problem with the current caching approach? What's involved in using an API key and what are the downsides?

Added an envvar `SYSL_GITHUB_TOKEN` to get user's GitHub PAT. And using `secrets.GITHUB_TOKEN` in CI as the access token.

ChloePlanet commented on pkg/mod/github.go:
```diff
@@ -0,0 +1,183 @@
+	path  string
+}
+
+func getGitHubRepoPath(filename string) (*githubRepoPath, error) {
```
Fixed by 2dac2800aced26eb30a0b3f7ce36d27be587d6c6

ChloePlanet commented on pkg/mod/github.go:
```diff
@@ -0,0 +1,183 @@
+	githubPath := filepath.Join(syslModulesCacheDir, "github.com")
+	githubDir, err := os.Open(githubPath)
+	if err != nil {
+		return err
```
Fixed by 9e541497adb602b775bf1a4d4c18f35d922d196e

ChloePlanet commented on .github/workflows/go.yml:
```diff
@@ -29,6 +29,7 @@ jobs:
         run: make coverage
         env:
           SYSL_PLANTUML: http://www.plantuml.com/plantuml
+          SYSL_GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Fixed by 4e6859faa2ba2b5cd4900954f49006b24c333a0d
